### PR TITLE
[1/7] lnwire: add new Gossip 1.75 messages

### PR DESCRIPTION
**Pull request by ellemouton**
This is the first step towards the [Gossip 1.75 epic](https://github.com/lightningnetwork/lnd/issues/7961)

In this PR, the definitions for the new gossip messages defined in the 
taproot-gossip proposal are implemented. This includes:
- `announcement_signatures_2`
- `channel_announcement_2`
- `channel_update_2`

Various updates are also made to some existing messages such as `channel_ready` and `gossip_timestamp_range`

I expect there to still be a few updates to these messages as the proposal progresses

### CODE REVIEW

File: lnwire/channel_update_2.go
Comment (Crypt-iQ) [inline lnwire/channel_update_2.go:259]: not sure if i'm off base here but the spec draft says it should be `MsgHash("channel_update_2", "bip340_sig", m)`?

File: lnwire/channel_ready.go
Comment (Crypt-iQ) [inline lnwire/channel_ready.go:24]: in the spec this one has a type value of 0 and the `ChanReadyAnnounceBtcNonce` has a val of 2

**Review by Crypt-iQ — COMMENTED** [review]
looks good, just a few questions

File: lnwire/channel_ready.go
Comment (Roasbeef) [inline lnwire/channel_ready.go:53]: Can be `fn.OptionalRecordT`. Then no need for the variables above for each type. Also this way, looking at the definition of the attribute, you already know what the TLV type will be. 

Once we merge https://github.com/lightningnetwork/lnd/pull/8334 then we can start to use it here.

File: lnwire/gossip_timestamp_range.go
Comment (Roasbeef) [inline lnwire/gossip_timestamp_range.go:49]: `tlv.OptionalRecordT[TlvType2, uint32]`

File: lnwire/gossip_timestamp_range.go
Comment (Roasbeef) [inline lnwire/gossip_timestamp_range.go:55]: `tlv.OptionalRecordT[TlvType4, uint32]`

File: lnwire/interfaces.go
Comment (Roasbeef) [inline lnwire/interfaces.go:10]: Would expect something around abstract sig validation as well? Or by the time we care about that, we've already done a type cast?

File: lnwire/interfaces.go
Comment (Roasbeef) [inline lnwire/interfaces.go:22]: Similarly, could also abstract validation of the chan ann behind this as well. Though at that point, maybe we'd move the interface somewhere else, or implement that as a superset interface.

File: lnwire/channel_update.go
Comment (Roasbeef) [inline lnwire/channel_update.go:333]: Perhaps let's try an enum value instead of -1, 0, 1, etc? Dunno about you, but I always find it challenging to read code that goes `if Compare() == -1`, vs `if Compare() == LessThan`

File: lnwire/msg_hash.go
Comment (Roasbeef) [inline lnwire/msg_hash.go:11]: Missing `godoc` comment.

File: lnwire/announcement_signatures_2.go
Comment (Roasbeef) [inline lnwire/announcement_signatures_2.go:62]: We've started to use the individual `Write` functions for stuff as it reduces allocs as the type cast isn't needed. See `WriteSatoshi`, etc.

File: lnwire/channel_announcement_2.go
Comment (Roasbeef) [inline lnwire/channel_announcement_2.go:88]: `tlv.OptionalRecordT`.

File: lnwire/boolean.go
Comment (Roasbeef) [inline lnwire/boolean.go:14]: Can just `tlv.RecordT[CustomType, bool]`

File: lnwire/channel_update_2.go
Comment (Roasbeef) [inline lnwire/channel_update_2.go:107]: `RecordT`

File: lnwire/channel_update_2.go
Comment (Roasbeef) [reply lnwire/channel_update_2.go:259]: yeah why would it sign ann sig and chan ann 2?

File: lnwire/channel_update_2.go
Comment (Roasbeef) [inline lnwire/channel_update_2.go:422]: Assuming `BlockHeight` is part of the interface, then we won't need to do a type cast here? This is basically the same as the `CmpAgge` on the other one, maybe we can: 
```go
func CmpAge[U ChannelUpdate1 | ChannelUpdate2](update1 U, update 2 U)
```

?

Dunno if would work out tho, I think you need to do a soft cast to be able to access teh shareed fields still

File: lnwire/channel_update_2.go
Comment (Roasbeef) [inline lnwire/channel_update_2.go:442]: Similarly these look essentially identical to the other one.

File: lnwire/node_announcement_2.go
Comment (Roasbeef) [inline lnwire/node_announcement_2.go:64]: `tlv.RecordT` errrrrwhere

**Review by Roasbeef — CHANGES_REQUESTED** [review]
Nice work! Need to cross ref with the spec, but the main comment is using `tlv.RecordT` in place of some of the new structs and helper functions.

File: discovery/gossiper_test.go
Comment (yyforyongyu) [inline discovery/gossiper_test.go:2360]: nit: wrong commit

File: lnwire/lnwire_test.go
Comment (yyforyongyu) [inline lnwire/lnwire_test.go:1128]: Feel like I've asked this before - what's the reason of doing this instead of, say, test both the cases deterministically?

File: lnwire/gossip_timestamp_range.go
Comment (yyforyongyu) [inline lnwire/gossip_timestamp_range.go:53]: Is this `ExtraData` ever used? Do we need migration?

File: lnwire/announcement_signatures_2.go
Comment (yyforyongyu) [inline lnwire/announcement_signatures_2.go:82]: Do we need a different msg type here? If so, then we are missing `MsgAnnounceSignatures1` and a few others

File: lnwire/announcement_signatures_2.go
Comment (yyforyongyu) [inline lnwire/announcement_signatures_2.go:11]:
```suggestion
// signatures by the sender to construct the channel_announcement message.
```

File: lnwire/announcement_signatures_2.go
Comment (yyforyongyu) [inline lnwire/announcement_signatures_2.go:8]: maybe mention who sends 1 and who sends 2? Atm the doc is just a dup of `AnnounceSignatures1`

File: lnwire/channel_announcement_2.go
Comment (yyforyongyu) [inline lnwire/channel_announcement_2.go:23]: I thought it's `announcement_sig`

File: lnwire/channel_announcement_2.go
Comment (yyforyongyu) [inline lnwire/channel_announcement_2.go:122]: What if we want to set it to something else, eg, in a regtest

File: lnwire/lnwire_test.go
Comment (yyforyongyu) [inline lnwire/lnwire_test.go:110]: 👍

File: lnwire/channel_announcement_2.go
Comment (yyforyongyu) [inline lnwire/channel_announcement_2.go:191]: Should `BitcoinKey1` and `BitcoinKey2` always be set or unset together?

File: lnwire/boolean.go
Comment (yyforyongyu) [inline lnwire/boolean.go:11]: Why do we choose to encode this way, seems complicated?

File: lnwire/boolean.go
Comment (yyforyongyu) [inline lnwire/boolean.go:27]: what if it's encoded explicitly?

**Review by yyforyongyu — COMMENTED** [review]
Nice work!🎙️ Finished initial round, mostly question, will read the specs in depth and come back 👀

File: lnwire/channel_ready.go
Comment (ellemouton) [reply lnwire/channel_ready.go]: fixed here :)

**Review by ellemouton — COMMENTED** [review]
Thanks for the initial reviews!

Updated. 

Main update here is that the validation logic for `ChannelAnnouncement` and `ChannelUpdate` has been moved to behind the new interfaces

File: lnwire/channel_announcement_2.go
Comment (yyforyongyu) [inline lnwire/channel_announcement_2.go:263]: Very nice pattern! I do have two concerns here,
- the fact that we need to pass `fetchPKScript` here implies `lnwire` is not equipped to run validation logic. It may look pure, while in essence it's still an extension of `gossiper`(also not in favor of passing functions as params or configs as it's a hack to fix circular imports). This means if things change in the upper layer, the lower layer also needs to be aware, not good for modularization. We may be tempted to fix it by importing methods, say, from `lnwallet`, which leads to my second concert,
- now that we have validators built for gossip messages, what about others? Should we also need to validate channel open or update htlc messages in `lnwire`?

So I would prefer keeping `lnwire` as pure as it is now - it defines wire messages, handles encoding/decoding bytes, and leave the rest to be handled by the upper layers.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Again very digestible commits💯 My main question is whether we should extend `lnwire` beyond encoding bytes.

File: channeldb/graph.go
Comment (ellemouton) [reply channeldb/graph.go:2861]: goland 🤓 "right click -> rename"

File: lnwire/gossip_timestamp_range.go
Comment (ellemouton) [reply lnwire/gossip_timestamp_range.go:44]: yeah i havent pushed up this change yet. Will do soon.

**Comment by ellemouton** [comment]
sorry about the lag here - it is on the list. Just need to get things through with the RB stuff first. Then focus is back on this

**Comment by ellemouton** [comment]
!lightninglabs-deploy mute

File: routing/router.go
Comment (ellemouton) [reply routing/router.go:1635]: agreeeeeed

**Review by ellemouton — COMMENTED** [review]
Ok, we're back in full swing!

This latest update:
1. has everything rebased on master (main complication were from the router->graph move)
2. removes the NodeAnnouncement2 addition and the rename of NodeAnnouncement1 as this first set of gossip PRs does not make use of it. 
3. Update given last review rounds and latest spec doc updates

### Main question & un-addressed review comment: 
One review comment in an earlier round suggested hiding validation of the announcement messages behind the new interfaces (which has now been done) but then a newer comment suggested that the `lnwire` package remain purely about wire encoding & should not contain validation. 

I agree with both comments so im wondering what you guys think is the best solution here?

File: lnwire/interfaces.go
Comment (ellemouton) [reply lnwire/interfaces.go:23]: ok have done this now but another review comment suggests that lnwire should not contain validation logic... what do you guys think is the best approach here? 

should we let the interfaces only describe underlying fields and then have separate interfaces somewhere else (netann) that hides the validation / sig check logic?

File: lnwire/boolean.go
Comment (ellemouton) [reply lnwire/boolean.go]: removed :)

File: lnwire/node_announcement_2.go
Comment (ellemouton) [reply lnwire/node_announcement_2.go]: removed node_announcement_2 for now :)

File: lnwire/channel_announcement.go
Comment (ellemouton) [inline lnwire/channel_announcement.go:195]: so @yyforyongyu  - carrying on from your suggestion re keeping the lnwire pkg purely about encoding & decoding wire messages, do you rate this validation should move to `netann` instead perhaps? 

Then we can maybe just have a single `Validate(ChannelAnnouncement)` function there with a switch on the underlying chan announce msg type?

File: lnwire/interfaces.go
Comment (ellemouton) [reply lnwire/interfaces.go:23]: or perhaps this interface stays here but we instead just keep the Validation functions separate in `netann`?

File: graph/ann_validation.go
Comment (ellemouton) [inline graph/ann_validation.go:124]: same thread as before: should we instead just extend these functions with switches on the types instead of moving the validation to the message interfaces?

**Comment by Roasbeef** [comment]
> I agree with both comments so im wondering what you guys think is the best solution here?

Perhaps the solution is moving the validation interface into the `netann` package?

File: lnwire/channel_announcement.go
Comment (Roasbeef) [reply lnwire/channel_announcement.go:195]: If we move the interface to the `netann` package, then we can have a wrapper struct the accepts either of the wire message types, using the field to implement the new abstracted validation.

Comment (Roasbeef) [reply lnwire/channel_announcement.go:195]: I think we should also make a type def for the accepted closure, which'll make the signature of the new interface easier to parse.

File: lnwire/interfaces.go
Comment (Roasbeef) [reply lnwire/interfaces.go:23]: > or perhaps this interface stays here but we instead just keep the Validation functions separate in netann?

+1

File: lnwire/announcement_signatures_2.go
Comment (Roasbeef) [inline lnwire/announcement_signatures_2.go:29]: Reloading context here, but don't we also need new nonces for the open+accept channel messages? Which are only set if the channel is to be announced.

File: lnwire/lnwire.go
Comment (Roasbeef) [inline lnwire/lnwire.go:945]: So we do already have an `Encode/Decode` method for `partialSig` that can be reused here.

File: lnwire/features.go
Comment (Roasbeef) [inline lnwire/features.go:647]: What about other instances that maybe were relying on this type? We should make sure to update those to avoid unintended breakage during deployment.

File: lnwire/announcement_signatures_2.go
Comment (Roasbeef) [reply lnwire/announcement_signatures_2.go:29]: Also: TLV errwhere?

File: lnwire/msat.go
Comment (Roasbeef) [inline lnwire/msat.go:59]: 👍

**Review by ellemouton — COMMENTED** [review]
Thanks @Roasbeef 🙏 

Updated with the interface changes we discussed 👍

File: channeldb/waitingproof.go
Comment (ellemouton) [reply channeldb/waitingproof.go:194]: cool - latest revision has this

File: lnwire/interfaces.go
Comment (ellemouton) [reply lnwire/interfaces.go:19]: updated!

File: lnwire/channel_announcement_2.go
Comment (ellemouton) [reply lnwire/channel_announcement_2.go]: updated 👍  lnwire now no longer contains the sign validation etc

**Review by Roasbeef — APPROVED** [review]
LGTM 🌊

Should be g2g after a linter fix: 
```
Error: lnwire/interfaces.go:58:1: directive `//nolint:interfacebloat` is unused for linter "interfacebloat" (nolintlint)
```

File: discovery/gossiper.go
Comment (yyforyongyu) [inline discovery/gossiper.go:1055]: 🙏

File: netann/msg_hash.go
Comment (yyforyongyu) [inline netann/msg_hash.go:1]: nit: typo in the commit msg `funcion` -> `function`

File: netann/channel_announcement.go
Comment (yyforyongyu) [inline netann/channel_announcement.go:252]: Q: what if only one of the keys is set, like `bitcoinKey1` is set but not `bitcoinKey2`?

File: lnwire/channel_update_2.go
Comment (yyforyongyu) [inline lnwire/channel_update_2.go:51]:
```suggestion
	// created and signed this message. If this field is present, it was
```

**Review by yyforyongyu — APPROVED** [review]
Thanks for the effort in writing all those tests🙏 I think it's good to go, just a few nits. Will refresh the specs and post questions in the following PRs, if any.

File: netann/channel_announcement.go
Comment (ellemouton) [reply netann/channel_announcement.go:252]: it will then fall to the `else` branch below which will fetch the on-chain script. 
ie, we dont explicitly need to not allow that case I dont think. It could maybe be a valid future message potentially.

**Comment by ellemouton** [comment]
Thanks for the review @Roasbeef  & @yyforyongyu  🔈

**Pull request merged**
